### PR TITLE
Example: Bayes NN with steinVI

### PR DESCRIPTION
**Pull request by OlaRonning**
See #1313. related to #833.

### CODE REVIEW

File: examples/stein_bnn.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    """
+
+    prec_nn = numpyro.sample(
+        "prec_nn", Gamma(1.0, 0.1)
+    )  # hyper prior for precision of nn weights and biases
+
+    n, m = x.shape
+
+    with numpyro.plate("l1_hidden", hidden_dim):
+        b1 = numpyro.sample(  # prior l1 bias term
```
Comment (fehiepsi) [inline examples/stein_bnn.py:67]: This is not formatted well

File: examples/stein_bnn.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            "nn_b1",
+            Normal(
+                0.0,
+                1.0 / prec_nn,
```
Comment (fehiepsi) [inline examples/stein_bnn.py:71]: Do you mean to take 1 / sqrt(precision)

File: examples/stein_bnn.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            ),
+        )
+        assert b1.shape == (hidden_dim,)
+
+        with numpyro.plate("l1_feat", m):
```
Comment (fehiepsi) [inline examples/stein_bnn.py:76]: It is a good practice to add dim to plate statements

File: examples/stein_bnn.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            w1 = numpyro.sample("nn_w1", Normal(0.0, 1.0 / prec_nn))  # prior l1 weights
+            assert w1.shape == (m, hidden_dim)
+
+    with numpyro.plate("l2_hidden", hidden_dim):
+        w2 = numpyro.sample("nn_w2", Normal(0.0, 1.0 / prec_nn))  # prior output weights
+
+    b2 = numpyro.sample("nn_b2", Normal(0.0, 1.0 / prec_nn))  # prior output bias term
```
Comment (fehiepsi) [inline examples/stein_bnn.py:83]: Ditto: sqrt(precision)?

File: examples/stein_bnn.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    prec_obs = numpyro.sample(  # precision prior on observations
+        "prec_obs", Gamma(1.0, 0.1)
+    )
+    with handlers.scale(scale=subsample_size), numpyro.plate(
```
Comment (fehiepsi) [inline examples/stein_bnn.py:88]: Could you comment on why we need to scale here?

File: numpyro/examples/datasets.py [hunk 6]
```diff
@@ -14,6 +14,7 @@
 import zipfile
 
 import numpy as np
+import scipy
```
Comment (fehiepsi) [inline numpyro/examples/datasets.py:17]: Could you move scipy to local function? We don't require scipy by default.

File: numpyro/examples/datasets.py [hunk 7]
```diff
@@ -70,6 +76,37 @@
     ["https://archive.ics.uci.edu/ml/machine-learning-databases/00280/HIGGS.csv.gz"],
 )
 
+LR_BANANA = dset(
+    "lr_banana", ["https://github.com/pyro-ppl/datasets/raw/master/benchmarks.mat"]
```
Comment (fehiepsi) [inline numpyro/examples/datasets.py:80]: The filename benchmarks is confusing. Could you rename the file? Also, if mat is not a convenient format (requires scipy) then it is better to store it as npz file

**Comment by fehiepsi** [comment]
@OlaRonning While you are working on this, could you expose SteinVI and its utility to docs? Currently it seems to be missing. It would be nice to write up a little docs for this module (see e.g. Pyro's [forecasting](https://docs.pyro.ai/en/stable/contrib.forecast.html) module) and points to examples that you add.

**Comment by OlaRonning** [comment]
I got a little sidetracked with another project.

> While working on this, could you expose SteinVI and its utility to docs? 

Sure, I'll expose the methods to docs in a separate PR and write module docs as well. 

> Then, I'll update the release docs for the 0.9.0 release.

Is the next release scheduled?

**Comment by fehiepsi** [comment]
Yes, I just made the release. Sorry about the SteinVI docs, I just recognized it missing while collecting the release notes. When it is available, I'll update the release notes to help readers easier to navigate. Thank you!

File: examples/stein_bnn.py (hunk 5 above)
Comment (OlaRonning) [reply examples/stein_bnn.py:88]: We don't need the scale. Performance is the same without scaling. If I recall, scaling mimics [temperature](https://docs.pymc.io/en/latest/api/inference.html#pymc.variational.inference.SVGD) in the pymc3 implementation.

File: examples/stein_bnn.py (hunk 3 above)
Comment (OlaRonning) [reply examples/stein_bnn.py:76]: It makes sense; I've updated with dim params set.

File: examples/stein_bnn.py (hunk 2 above)
Comment (OlaRonning) [reply examples/stein_bnn.py:71]: yeah, ups

File: numpyro/examples/datasets.py (hunk 7 above)
Comment (OlaRonning) [reply numpyro/examples/datasets.py:80]: I'll fix the names and convert them to NumPy format.

File: numpyro/examples/datasets.py (hunk 6 above)
Comment (OlaRonning) [reply numpyro/examples/datasets.py:17]: I'll convert the format as you suggest below and remove scipy.

File: numpyro/examples/datasets.py (hunk 7 above)
Comment (OlaRonning) [reply numpyro/examples/datasets.py:80]: On second thought, I removed it. I'm not sure I'll add the example to numpyro, and if I do, I'll fix the problem then.

**Comment by OlaRonning** [comment]
closed to fix rebase!

**Comment by fehiepsi** [comment]
@OlaRonning it looks like black just released its stable version and changed x ** 2 into x**2 . Could you update black and run `make format`?

**Comment by OlaRonning** [comment]
That makes sense. I'll fix it in a sec.

**Review by fehiepsi — APPROVED** [review]
Thanks, @OlaRonning! The example is easy to follow 💯

**Pull request merged**